core/src/main/java/net/opentsdb/query/processor/bucketquantile/BucketQuantileConfig.java

```diff
@@ -0,0 +1,618 @@
+  }
+
+  public double getNanThreshold() {
+    return nan_threshold;
+  }
+  
+  public double getMissingMetricThreshold() {
+    return missing_metric_threshold;
+  }
+  
+  /** @return Whether or not NaNs should be treated as sentinels or considered 
+   * in arithmetic. */
+  public boolean getInfectiousNan() {
+    return infectious_nan;
+  }
+
+  @Override
+  public boolean pushDown() {
+    return false;
+  }
+
+  @Override
+  public boolean joins() {
+    return true;
+  }
+
+  @Override
+  public Builder toBuilder() {
+    final Builder builder = new Builder()
+        .setAs(as)
+        .setBucketRegex(bucket_regex)
+        .setOverflowMax(overflow_max)
+        .setOverflow(overflow)
+        .setOverflowMetric(overflow_metric)
+        .setUnderflowMin(underflow_min)
+        .setOverflowId(overflow_id)
+        .setUnderflow(underflow)
+        .setUnderflowMetric(underflow_metric)
+        .setUnderflowId(underflow_id)
+        .setOutputOfBucket(output_of_bucket)
+        .setHistograms(Lists.newArrayList(histograms))
+        .setHistogramMetrics(histogram_metrics == null ? null : 
+            Lists.newArrayList(histogram_metrics))
+        .setHistogramIds(histogram_ids == null ? null : 
+            Lists.newArrayList(histogram_ids))
+        .setInfectiousNan(infectious_nan)
+        .setQuantiles(Lists.newArrayList(quantiles))
+        .setCumulativeBuckets(cumulative_buckets)
+        .setCounterBuckets(counter_buckets)
+        .setNanThreshold(nan_threshold)
+        .setMissingMetricThreshold(missing_metric_threshold);
+    super.toBuilder(builder);
+    return builder;
+  }
+  
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    final BucketQuantileConfig config = (BucketQuantileConfig) o;
+    return Objects.equal(as, config.getAs()) &&
+        Objects.equal(bucket_regex, config.bucket_regex) &&
+        Objects.equal(overflow, config.overflow) &&
+        Objects.equal(overflow_max, config.overflow_max) &&
+        Objects.equal(underflow, config.underflow) &&
+        Objects.equal(underflow_min, config.underflow_min) &&
+        Objects.equal(output_of_bucket, config.output_of_bucket) &&
+        Objects.equal(histograms, config.getHistograms()) &&
+        Objects.equal(quantiles, config.quantiles) &&
+        Objects.equal(cumulative_buckets, config.cumulative_buckets) &&
+        Objects.equal(counter_buckets, config.counter_buckets) &&
+        Objects.equal(nan_threshold, config.nan_threshold) &&
+        Objects.equal(missing_metric_threshold, config.missing_metric_threshold) &&
+        Objects.equal(infectious_nan, config.getInfectiousNan()) &&
+        Objects.equal(interpolator_configs, config.interpolator_configs) &&
+        Objects.equal(id, config.getId());
+  }
+  
+  @Override
+  public int hashCode() {
+    return buildHashCode().asInt();
```

Comment:
Since all the fields are final for this class, it makes sense to compute this hashcode and store once instead of calculating the expensive buildHashCode() call every time a comparator call asks for a hashCode.

Reply:
Ack. I have caching in most locations, I need to add it here.